--- ./bin/qpeek.orig	2007-09-14 14:22:56.737346238 +0200
+++ ./bin/qpeek	2007-09-14 14:25:34.125299051 +0200
@@ -35,6 +35,7 @@
 	       "/var/spool/batch/torque",
 	       "/var/spool/batch/pbs-piv",
 	       "/var/spool/batch/pbs-ipf",
+	       "/opt/torque",
 	       ".",
 	      );
     foreach $dir ( @defaults )
@@ -139,7 +140,7 @@
 # chop to magic pbs length
 $jobname = "$jobid.$pbsserver";
 $jobname =~ s/(.{11}).*/$1/;
-exec "rsh -n $node $tool $numlines $spool/spool/$jobname.$suffix\n";
+exec "ssh -n $node $tool $numlines $spool/spool/$jobname.$suffix\n";
 
 sub mothersuperior
 {
@@ -153,7 +154,7 @@
       if ( $_ =~ /exec_host/ )
         {
 	  ($keyword,$node)=split(/=/);
-	  $node=~s:/[0-9]+[A-z0-9/+-]*::;
+	  $node=~s:/[0-9]+[A-z0-9/+.-]*::;
 	  $keyword="";
 	}
     }
